File: backend/src/main/java/com/example/realestate/model/Project.java
package com.example.realestate.model;

import com.fasterxml.jackson.annotation.JsonAlias;
import jakarta.persistence.*;

@Entity
@Table(name = "projects")
public class Project {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "owner_id")
    private Long ownerId;

    @JsonAlias({"landId", "land_id"})
    @Column(name = "land_id")
    private Long landId;

    @JsonAlias({"projectName", "title"})
    @Column(name = "project_name", nullable = false)
    private String projectName;

    @Column(nullable = false)
    private String location;

    @JsonAlias({"landSize", "land_size"})
    @Column(name = "land_size", nullable = false)
    private double landSize = 0.0;

    @JsonAlias({"investmentRequired", "capitalRequired", "capital_required"})
    @Column(name = "investment_required", nullable = false)
    private double investmentRequired = 0.0;

    @JsonAlias({"expectedROI", "projectedGrowth", "expected_roi"})
    @Column(name = "expected_roi", nullable = false)
    private double expectedROI = 0.0;

    @JsonAlias({"expectedIRR", "irr", "expected_irr"})
    @Column(name = "expected_irr", nullable = false)
    private double expectedIRR = 0.0;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ProjectStage stage;

    // Tourism classification
    @Column(name = "state_category")
    private String stateCategory;

    private String destination;

    // Investment analytics (tourism real estate)
    @JsonAlias({"rentalYield", "rental_yield"})
    @Column(name = "rental_yield")
    private double rentalYield = 0.0;

    @JsonAlias({"occupancyRate", "occupancy_rate"})
    @Column(name = "occupancy_rate")
    private double occupancyRate = 0.0;

    @JsonAlias({"estimatedMonthlyIncome", "estimated_monthly_income"})
    @Column(name = "estimated_monthly_income")
    private Double estimatedMonthlyIncome;

    @JsonAlias({"projectedAnnualIncome", "projected_annual_income"})
    @Column(name = "projected_annual_income")
    private double projectedAnnualIncome = 0.0;

    @JsonAlias({"breakEvenYears", "break_even_years"})
    @Column(name = "break_even_years")
    private double breakEvenYears = 0.0;

    @JsonAlias({"capitalAppreciation5Year", "capital_appreciation_5y"})
    @Column(name = "capital_appreciation_5y")
    private double capitalAppreciation5Year = 0.0;

    @JsonAlias({"peakSeasonOccupancy", "peak_season_occupancy"})
    @Column(name = "peak_season_occupancy")
    private double peakSeasonOccupancy = 0.0;

    @JsonAlias({"seasonalDemand", "seasonal_demand"})
    @Column(name = "seasonal_demand")
    private String seasonalDemand = "MEDIUM";

    @JsonAlias({"averageDailyRate", "average_daily_rate", "adr"})
    @Column(name = "average_daily_rate")
    private double averageDailyRate = 0.0;

    @JsonAlias({"monthlyCashFlow", "monthly_cash_flow"})
    @Column(name = "monthly_cash_flow")
    private double monthlyCashFlow = 0.0;

    @JsonAlias({"netOperatingIncome", "net_operating_income", "noi"})
    @Column(name = "net_operating_income")
    private double netOperatingIncome = 0.0;

    // NEW ANALYTICS FIELDS
    private Double capitalAppreciation;
    private Double averageOccupancy;
    private Double peakOccupancy;
    private Double adr;
    private Double noi;

    public Project() {
    }

    @PrePersist
    public void ensureDefaultStage() {
        if (this.stage == null) this.stage = ProjectStage.LAND_APPROVED;
        if (this.seasonalDemand == null || this.seasonalDemand.isBlank()) {
            this.seasonalDemand = "MEDIUM";
        }
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(Long ownerId) {
        this.ownerId = ownerId;
    }

    public Long getLandId() {
        return landId;
    }

    public void setLandId(Long landId) {
        this.landId = landId;
    }

    public String getProjectName() {
        return projectName;
    }

    public void setProjectName(String projectName) {
        this.projectName = projectName;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public double getLandSize() {
        return landSize;
    }

    public void setLandSize(double landSize) {
        this.landSize = landSize;
    }

    public double getInvestmentRequired() {
        return investmentRequired;
    }

    public void setInvestmentRequired(double investmentRequired) {
        this.investmentRequired = investmentRequired;
    }

    public double getExpectedROI() {
        return expectedROI;
    }

    public void setExpectedROI(double expectedROI) {
        this.expectedROI = expectedROI;
    }

    public double getExpectedIRR() {
        return expectedIRR;
    }

    public void setExpectedIRR(double expectedIRR) {
        this.expectedIRR = expectedIRR;
    }

    public ProjectStage getStage() {
        return stage;
    }

    public void setStage(ProjectStage stage) {
        this.stage = stage;
    }

    public String getStateCategory() {
        return stateCategory;
    }

    public void setStateCategory(String stateCategory) {
        this.stateCategory = stateCategory;
    }

    public String getDestination() {
        return destination;
    }

    public void setDestination(String destination) {
        this.destination = destination;
    }

    public double getRentalYield() {
        return rentalYield;
    }

    public void setRentalYield(double rentalYield) {
        this.rentalYield = rentalYield;
    }

    public double getOccupancyRate() {
        return occupancyRate;
    }

    public void setOccupancyRate(double occupancyRate) {
        this.occupancyRate = occupancyRate;
    }

    public Double getEstimatedMonthlyIncome() {
        return estimatedMonthlyIncome;
    }

    public void setEstimatedMonthlyIncome(Double estimatedMonthlyIncome) {
        this.estimatedMonthlyIncome = estimatedMonthlyIncome;
    }

    public double getProjectedAnnualIncome() {
        return projectedAnnualIncome;
    }

    public void setProjectedAnnualIncome(double projectedAnnualIncome) {
        this.projectedAnnualIncome = projectedAnnualIncome;
    }

    public double getBreakEvenYears() {
        return breakEvenYears;
    }

    public void setBreakEvenYears(double breakEvenYears) {
        this.breakEvenYears = breakEvenYears;
    }

    public double getCapitalAppreciation5Year() {
        return capitalAppreciation5Year;
    }

    public void setCapitalAppreciation5Year(double capitalAppreciation5Year) {
        this.capitalAppreciation5Year = capitalAppreciation5Year;
    }

    public double getPeakSeasonOccupancy() {
        return peakSeasonOccupancy;
    }

    public void setPeakSeasonOccupancy(double peakSeasonOccupancy) {
        this.peakSeasonOccupancy = peakSeasonOccupancy;
    }

    public String getSeasonalDemand() {
        return seasonalDemand;
    }

    public void setSeasonalDemand(String seasonalDemand) {
        this.seasonalDemand = seasonalDemand;
    }

    public double getAverageDailyRate() {
        return averageDailyRate;
    }

    public void setAverageDailyRate(double averageDailyRate) {
        this.averageDailyRate = averageDailyRate;
    }

    public double getMonthlyCashFlow() {
        return monthlyCashFlow;
    }

    public void setMonthlyCashFlow(double monthlyCashFlow) {
        this.monthlyCashFlow = monthlyCashFlow;
    }

    public double getNetOperatingIncome() {
        return netOperatingIncome;
    }

    public void setNetOperatingIncome(double netOperatingIncome) {
        this.netOperatingIncome = netOperatingIncome;
    }

    public Double getCapitalAppreciation() {
        return capitalAppreciation;
    }

    public void setCapitalAppreciation(Double capitalAppreciation) {
        this.capitalAppreciation = capitalAppreciation;
    }

    public Double getAverageOccupancy() {
        return averageOccupancy;
    }

    public void setAverageOccupancy(Double averageOccupancy) {
        this.averageOccupancy = averageOccupancy;
    }

    public Double getPeakOccupancy() {
        return peakOccupancy;
    }

    public void setPeakOccupancy(Double peakOccupancy) {
        this.peakOccupancy = peakOccupancy;
    }

    public Double getAdr() {
        return adr;
    }

    public void setAdr(Double adr) {
        this.adr = adr;
    }

    public Double getNoi() {
        return noi;
    }

    public void setNoi(Double noi) {
        this.noi = noi;
    }
}

File: lib/models/project.dart
class Project {
  final int? id;
  final int? landId;
  final int? ownerId; // New field for ownership
  final String projectName;
  final String location;
  final double landSize;
  final double investmentRequired;
  final double expectedROI;
  final double expectedIRR;
  final String stage;
  final String? stateCategory;
  final String? destination;

  // Investment metrics
  final double? rentalYield;
  final double? projectedAnnualIncome;
  final double? capitalAppreciation;
  final double breakEvenYears;

  // Demand metrics
  final double occupancyRate;
  final double? averageOccupancy;
  final double? peakOccupancy;
  final String? seasonalDemand;
  final double averageDailyRate;
  final double? adr;

  // Financial metrics
  final double monthlyCashFlow;
  final double netOperatingIncome;
  final double? noi;

  // Compatibility fields
  final double? estimatedMonthlyIncome;
  final double capitalAppreciation5Year;
  final double peakSeasonOccupancy;

  Project({
    this.id,
    this.landId,
    this.ownerId,
    required this.projectName,
    required this.location,
    this.landSize = 0.0,
    this.investmentRequired = 0.0,
    this.expectedROI = 0.0,
    this.expectedIRR = 0.0,
    this.stage = 'LAND_APPROVED',
    this.stateCategory,
    this.destination,
    this.rentalYield,
    this.projectedAnnualIncome,
    this.capitalAppreciation,
    this.breakEvenYears = 0.0,
    this.occupancyRate = 0.0,
    this.averageOccupancy,
    this.peakOccupancy,
    this.seasonalDemand,
    this.averageDailyRate = 0.0,
    this.adr,
    this.monthlyCashFlow = 0.0,
    this.netOperatingIncome = 0.0,
    this.noi,
    this.estimatedMonthlyIncome,
    this.capitalAppreciation5Year = 0.0,
    this.peakSeasonOccupancy = 0.0,
  });

  String get title => projectName;

  factory Project.fromJson(Map<String, dynamic> json) {
    final projectName = (json['projectName'] ?? json['title'] ?? '') as String;
    final location = (json['location'] ?? '') as String;

    double parseDouble(dynamic v) {
      if (v == null) return 0.0;
      if (v is num) return v.toDouble();
      return double.tryParse(v.toString()) ?? 0.0;
    }

    double? parseDoubleOpt(dynamic v) {
      if (v == null) return null;
      if (v is num) return v.toDouble();
      return double.tryParse(v.toString());
    }

    return Project(
      id: json['id'] as int?,
      landId: json['landId'] ?? json['land_id'],
      ownerId: json['ownerId'] ?? json['owner_id'],
      projectName: projectName,
      location: location,
      landSize: parseDouble(json['landSize'] ?? json['land_size']),
      investmentRequired: parseDouble(json['investmentRequired'] ?? json['capital_required']),
      expectedROI: parseDouble(json['expectedROI'] ?? json['expected_roi']),
      expectedIRR: parseDouble(json['expectedIRR'] ?? json['expected_irr']),
      stage: (json['stage'] ?? 'LAND_APPROVED') as String,
      stateCategory: json['stateCategory'] ?? json['state_category'],
      destination: json['destination'],
      rentalYield: parseDoubleOpt(json['rentalYield']),
      projectedAnnualIncome: parseDoubleOpt(json['projectedAnnualIncome']),
      capitalAppreciation: parseDoubleOpt(json['capitalAppreciation']),
      breakEvenYears: parseDouble(json['breakEvenYears'] ?? json['break_even_years']),
      occupancyRate: parseDouble(json['occupancyRate'] ?? json['occupancy_rate']),
      averageOccupancy: parseDoubleOpt(json['averageOccupancy']),
      peakOccupancy: parseDoubleOpt(json['peakOccupancy']),
      seasonalDemand: json['seasonalDemand'],
      averageDailyRate: parseDouble(json['averageDailyRate'] ?? json['average_daily_rate']),
      adr: parseDoubleOpt(json['adr']),
      monthlyCashFlow: parseDouble(json['monthlyCashFlow'] ?? json['monthly_cash_flow']),
      netOperatingIncome: parseDouble(json['netOperatingIncome'] ?? json['net_operating_income']),
      noi: parseDoubleOpt(json['noi']),
      estimatedMonthlyIncome: parseDoubleOpt(json['estimatedMonthlyIncome']),
      capitalAppreciation5Year: parseDouble(json['capitalAppreciation5Year'] ?? 0.0),
      peakSeasonOccupancy: parseDouble(json['peakSeasonOccupancy'] ?? 0.0),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (landId != null) 'landId': landId,
      if (ownerId != null) 'ownerId': ownerId,
      'projectName': projectName,
      'location': location,
      'landSize': landSize,
      'investmentRequired': investmentRequired,
      'expectedROI': expectedROI,
      'expectedIRR': expectedIRR,
      'stage': stage,
      if (stateCategory != null) 'stateCategory': stateCategory,
      if (destination != null) 'destination': destination,
      'rentalYield': rentalYield,
      'projectedAnnualIncome': projectedAnnualIncome,
      'capitalAppreciation': capitalAppreciation,
      'breakEvenYears': breakEvenYears,
      'occupancyRate': occupancyRate,
      'averageOccupancy': averageOccupancy,
      'peakOccupancy': peakOccupancy,
      'seasonalDemand': seasonalDemand,
      'averageDailyRate': averageDailyRate,
      'adr': adr,
      'monthlyCashFlow': monthlyCashFlow,
      'netOperatingIncome': netOperatingIncome,
      'noi': noi,
      'estimatedMonthlyIncome': estimatedMonthlyIncome,
      'capitalAppreciation5Year': capitalAppreciation5Year,
      'peakSeasonOccupancy': peakSeasonOccupancy,
    };
  }

  // Compatibility helpers
  String get theme => 'General';
  String get description => '';
  double get projectedGrowth => expectedROI;
  int get demandIndex => 5;
  String get riskProfile => 'Medium';
  double get irr => expectedIRR;
  double get capitalRequired => investmentRequired;
  double get capitalRaised => 0.0;
}
